src/libsyntax/ext/expand.rs

```diff
@@ -308,6 +311,50 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
                 items.push(item);
                 kind.expect_from_annotatables(items)
             }
+            SyntaxExtension::AttrProcMacro(ref mac) => {
+                let attr_toks = TokenStream::from_tts(tts_for_attr(&attr, &self.cx.parse_sess));
+                let item_toks = TokenStream::from_tts(tts_for_item(&item, &self.cx.parse_sess));
+
+                let tok_result = mac.expand(self.cx, attr.span, attr_toks, item_toks);
+                let parser = self.cx.new_parser_from_tts(&tok_result.to_tts());
+                let result = Box::new(TokResult { parser: parser, span: attr.span });
+
+                match item {
+                    Annotatable::Item(_) => {
```

Comment:
One last simplification that I didn't notice earlier -- I believe this whole match can be replaced with
```rust
kind.make_from(result).unwrap_or_else(|| {
    let msg = format!("macro could not be expanded into {} position", kind.name());
    self.cx.span_err(&msg);
    kind.dummy(attr.span)
})